### ISSUE TITLE
Timeout error for restore?

### ISSUE THREAD

**Issue by jamessewell** [pgbackrest/pgbackrest#848]
PGBackrest 2.17 from RPM on OEL 7.4

After ~4.5 hours  of restoring we get this - which looks like some sort of timeout? This is the second time this restore has failed like this.

```
2019-09-25 17:46:42.021 P00  TRACE:         Protocol::Base::Master->outputRead: strProtocolResult = {"err":42,"errStack":"protocol/client:protocolClientReadOutput:180:(trace log level required for parameters)\nprotocol/client:protocolCli
entExecute:(trace log level required for parameters)\nstorage/remote/read:storageReadRemoteOpen:(trace log level required for parameters)\ncommon/io/read:ioReadOpen:(trace log level required for parameters)\nstorage/storage:storageCopy:(
debug log level required for parameters)\ncommand/restore/file:restoreFile:(debug log level required for parameters)\ncommand/restore/protocol:restoreProtocol:(debug log level required for parameters)\nprotocol/server:protocolServerProce
ss:(debug log level required for parameters)\ncommand/local/local:cmdLocal:(debug log level required for parameters)\nmain:main:(debug log level required for parameters)","out":"raised from remote-20 protocol on 'rprnab001-bck.domain.prd
.int': unable to read data from remote-20 after 1830000ms"}
2019-09-25 17:46:42.021 P00  ERROR: [042]: raised from local-20 process: raised from remote-20 protocol on 'rprnab001-bck.domain.prd.int': unable to read data from remote-20 after 1830000ms
                                        protocol/client:protocolClientReadOutput:180:(trace log level required for parameters)
                                        protocol/client:protocolClientExecute:(trace log level required for parameters)
                                        storage/remote/read:storageReadRemoteOpen:(trace log level required for parameters)
                                        common/io/read:ioReadOpen:(trace log level required for parameters)
                                        storage/storage:storageCopy:(debug log level required for parameters)
                                        command/restore/file:restoreFile:(debug log level required for parameters)
                                        command/restore/protocol:restoreProtocol:(debug log level required for parameters)
                                        protocol/server:protocolServerProcess:(debug log level required for parameters)
                                        command/local/local:cmdLocal:(debug log level required for parameters)
                                        main:main:(debug log level required for parameters)
 at /loader/0x217ced8/pgBackRest/Protocol/Base/Master.pm line 173.
                                        pgBackRest::Protocol::Base::Master::outputRead(undef, 1, undef, undef, 1) called at /loader/0x217ced8/pgBackRest/Protocol/Local/Process.pm line 332
                                        eval {...} called at /loader/0x217ced8/pgBackRest/Protocol/Local/Process.pm line 345
                                        pgBackRest::Protocol::Local::Process::process('pgBackRest::Protocol::Local::Process=HASH(0x2c5fb38)') called at /loader/0x217ced8/pgBackRest/Restore.pm line 1305
                                        pgBackRest::Restore::process('pgBackRest::Restore=HASH(0x2b29cd0)') called at /loader/0x217ced8/pgBackRest/Main.pm line 153
                                        eval {...} called at /loader/0x217ced8/pgBackRest/Main.pm line 193
                                        pgBackRest::Main::main('restore') called at (eval 279) line 1
                                        eval '($iResult, $bErrorC, $strMessage) = pgBackRest::Main::main(\'restore\')
                                   ;' called at (eval 279) line 0
                                   2019-09-25 17:46:42.022 P00  ERROR: [042]: raised from local-20 process: raised from remote-20 protocol on 'rprnab001-bck.domain.prd.int': unable to read data from remote-20 after 1830000ms
                                        protocol/client:protocolClientReadOutput:180:(trace log level required for parameters)
                                        protocol/client:protocolClientExecute:(trace log level required for parameters)
                                        storage/remote/read:storageReadRemoteOpen:(trace log level required for parameters)
                                        common/io/read:ioReadOpen:(trace log level required for parameters)
                                        storage/storage:storageCopy:(debug log level required for parameters)
                                        command/restore/file:restoreFile:(debug log level required for parameters)
                                        command/restore/protocol:restoreProtocol:(debug log level required for parameters)
                                        protocol/server:protocolServerProcess:(debug log level required for parameters)
                                        command/local/local:cmdLocal:(debug log level required for parameters)
                                        main:main:(debug log level required for parameters)
 at /loader/0x217ced8/pgBackRest/Common/Log.pm line 595.
                                        pgBackRest::Common::Log::logException('pgBackRest::Common::Exception=HASH(0x11d8a5b0)') called at /loader/0x217ced8/pgBackRest/Protocol/Local/Process.pm line 358
                                        pgBackRest::Protocol::Local::Process::process('pgBackRest::Protocol::Local::Process=HASH(0x2c5fb38)') called at /loader/0x217ced8/pgBackRest/Restore.pm line 1305
                                        pgBackRest::Restore::process('pgBackRest::Restore=HASH(0x2b29cd0)') called at /loader/0x217ced8/pgBackRest/Main.pm line 153
                                        eval {...} called at /loader/0x217ced8/pgBackRest/Main.pm line 193
                                        pgBackRest::Main::main('restore') called at (eval 279) line 1
                                        eval '($iResult, $bErrorC, $strMessage) = pgBackRest::Main::main(\'restore\')
                                   ;' called at (eval 279) line 0
                                   2019-09-25 17:46:42.428 P00  TRACE:         Protocol::Command::Master->close(): bComplete = <false>

111

**Comment by jamessewell** [comment]
We did an incremental backup to try and correct any data errors (although I'm not sure if the above could be caused by that) and got the same error.

So it's happened three times now - any ideas?

**Comment by dwsteele** [comment]
The C migration of the restore command will be released tomorrow.  While this timeout does appear to be in the C portion of the restore worker code, we'll want to see it happen in 2.18 since about 95% of the code has been rewritten.

Having said that, if you are able to run your restore with `--log-subprocess` we might be able to get some information about what the remote is doing when the timeout happens.

**Comment by jamessewell** [comment]
Upgraded, but this is still happening.

```
-------------------PROCESS START-------------------
2019-10-07 16:57:28.979 P00   INFO: restore command begin 2.18: --delta --link-all --pg1-path=/ppas/9.6/data/pg_data --process-max=30 --repo1-host=r001-bck.domain.prd.int --repo1-host-user=napback --stanza=nap-planning
2019-10-07 16:57:42.573 P00   INFO: restore backup set 20190920-142605F_20191005-204255I
2019-10-07 16:57:42.576 P00   INFO: remove invalid files/links/paths from '/ppas/9.6/data/pg_data'
2019-10-07 16:57:52.041 P00   INFO: remove invalid files/links/paths from '/ppas/9.6/data/pg_xlog/'
2019-10-07 16:58:59.612 P02   INFO: restore file /ppas/9.6/data/pg_data/base/77485/8988528 (1GB, 0%) checksum 6c29c33f42a7b183f19132ea377c22ce6d829479
2019-10-07 16:59:09.193 P19   INFO: restore file /ppas/9.6/data/pg_data/base/77485/8984057.3 (1GB, 0%) checksum 524d331106facf57ad3a0fc47473e9d100a6b14f
...
2019-10-07 18:24:29.765 P26   INFO: restore file /ppas/9.6/data/pg_data/base/77485/8979287.13 (1GB, 3%) checksum a77ab92f5261b0c421d3414c2a82722192a62a36
2019-10-07 18:55:25.719 P00  ERROR: [042]: raised from local-3 protocol: raised from remote-3 protocol on 'r001-bck.domain.prd.int': unable to read data from remote-3 after 1830000ms
2019-10-07 18:55:41.873 P00   INFO: restore command end: aborted with exception [042]
```

I will get the `--log-subprocess` output now.

**Comment by jamessewell** [comment]
I have a full trace but it's large (and I'd rather not post it publicly). I can provide a link by email if that helps?

**Comment by dwsteele** [comment]
So here's the error from the log you sent:
```
2019-10-08 14:12:03.022 P00  ERROR: [042]: raised from local-5 protocol: raised from remote-5 protocol on 'XXX': unable to read data from remote-5 after 3600000ms
```
But process 5 had just reported results here:
```
2019-10-08 14:11:57.249 P05 DETAIL: restore file <...>
```
That's a lot less than 3600 seconds so it looks like the idle timer is not being reset.  We'll have a that.

**Comment by jamessewell** [comment]
Thanks!

Would we be able to restore this using another version? We had previously
(before 2.17) been on 2.05 and that’s worked in the past.

Or has the backup format changed?

Cheers,
James Sewell

**Comment by dwsteele** [comment]
I think the easiest thing would be to set `protocol-timeout` higher than your expected restore time.  The repo format has not changed since 1.00 but going all the way back to 2.05 will just give you other baggage.

The current problem is that a keep alive is not being sent to the remote while the restore is checksumming local files to determine what can be kept and what needs to be fetched from the remote repo.  The good news is that this is a pretty simple fix.

**Comment by Tostino** [comment]
Just wanted to throw in that I got hit by this last week when trying to do a delta restore to a small VM with way less IO than production, restore took forever to checksum files and I got the same timeout.

Glad it's already known.

**Comment by jamessewell** [comment]
Would setting it too high cause any issues? The backup takes ~50 hours so I
assume the restore could take around that time (although I can give it 3x
the CPU).

Thanks again,
James Sewell

**Comment by dwsteele** [comment]
The main drawback of setting protocol-timeout higher is that it will take longer to discover a legitimate network issue.  Other than that, nothing that we know of.

The restore timeout will not happen as long as files are being copied from the repo.  The problem happens when a long time is spent doing delta checksums.  I imagine this should be at least an order of magnitude faster than your backups so `protocol-timeout=18000` should be safe.

**Comment by dwsteele** [comment]
Fixed at 47e0ba2502e5c3e8eb0.
